Include/internal/pycore_bytesobject.h

```diff
@@ -33,6 +34,56 @@ _PyBytes_ReverseFind(const char *haystack, Py_ssize_t len_haystack,
                      const char *needle, Py_ssize_t len_needle,
                      Py_ssize_t offset);
 
+
+/** Helper function to implement the inplace repeat method on a buffer
+ *
+ * len_dest is assumed to be an integer multiple of len_src.
+ *
+ * This method repeately doubles the number of bytes copied to reduce
+ * the number of invocations of memcpy
+ */
+static inline void
+_PyBytes_RepeatInPlace(char* buffer, Py_ssize_t start_len, Py_ssize_t end_len)
+{
+    if (start_len == 1)
+        memset(buffer, buffer[0], end_len);
+    else {
+        Py_ssize_t copied = start_len;
+        while (copied < end_len) {
+            Py_ssize_t bytes_to_copy = Py_MIN(copied, end_len - copied);
+            memcpy(buffer + copied, buffer, bytes_to_copy);
+            copied += bytes_to_copy;
+        }
+    }
+}
+
+/** Helper function to implement the repeat method on a buffer
+ *
+ * len_dest is assumed to be an integer multiple of len_src.
+ * 
+ * This method repeately doubles the number of bytes copied to reduce
+ * the number of invocations of memcpy.
+ */
+static inline void
+_PyBytes_Repeat(char* dest, Py_ssize_t len_dest,
+                const char* src, Py_ssize_t len_src)
+{
+    if (len_dest > 0) {
+        if (len_src == 1)
+            memset(dest, src[0], len_dest);
+        else {
+            memcpy(dest, src, len_src);
+
+            Py_ssize_t copied = len_src;
+            while (copied < len_dest) {
+                Py_ssize_t bytes_to_copy = Py_MIN(copied, len_dest - copied);
+                memcpy(dest + copied, dest, bytes_to_copy);
+                copied += bytes_to_copy;
+            }
+        }
+    }
```

Comment:
```suggestion
    if (len_dest > 0) {
        memcpy(dest, src, len_src);
        _PyBytes_RepeatInPlace(dest, len_src, len_dest);     
   }
```

Reply:
@sweeneyde Thanks for the suggestion. If really simplifies the code. Part of this I already considered, but did not push because it seems to hurt the small cases. E.g. `[None] * 10` or `b'a' * 5`.

Both your suggestions (e.g. no inlining and skipping the special case where `len_src==1`) I implemented in https://github.com/eendebakpt/cpython/tree/performance/array_inplace_repeat_v2

Microbenchmarks show an overall slowdown, in particular for the cases with a low number of repeats or length of the object equal to 1.  
```
array repeat 1: Mean +- std dev: [patchfull] 185 ns +- 9 ns -> [patchv2full] 191 ns +- 1 ns: 1.03x slower
array repeat 2: Mean +- std dev: [patchfull] 185 ns +- 2 ns -> [patchv2full] 194 ns +- 2 ns: 1.05x slower
array repeat 10: Mean +- std dev: [patchfull] 211 ns +- 5 ns -> [patchv2full] 220 ns +- 3 ns: 1.05x slower
array repeat 1000: Mean +- std dev: [patchfull] 1.67 us +- 0.01 us -> [patchv2full] 1.70 us +- 0.01 us: 1.02x slower
array repeat 5000: Mean +- std dev: [patchfull] 9.18 us +- 0.07 us -> [patchv2full] 9.43 us +- 0.08 us: 1.03x slower
bytes(1) repeat 2: Mean +- std dev: [patchfull] 48.3 ns +- 0.7 ns -> [patchv2full] 50.7 ns +- 0.5 ns: 1.05x slower
bytes(1) repeat 10: Mean +- std dev: [patchfull] 48.1 ns +- 3.0 ns -> [patchv2full] 49.4 ns +- 0.6 ns: 1.03x slower
bytes(1) repeat 100: Mean +- std dev: [patchfull] 47.6 ns +- 0.5 ns -> [patchv2full] 50.2 ns +- 0.9 ns: 1.05x slower
bytes(1) repeat 1000: Mean +- std dev: [patchfull] 99.8 ns +- 0.5 ns -> [patchv2full] 103 ns +- 1 ns: 1.03x slower
array(1) repeat 1: Mean +- std dev: [patchfull] 84.4 ns +- 0.5 ns -> [patchv2full] 89.1 ns +- 1.7 ns: 1.06x slower
array(1) repeat 10: Mean +- std dev: [patchfull] 96.3 ns +- 1.3 ns -> [patchv2full] 101 ns +- 1 ns: 1.05x slower
array(1) repeat 1000: Mean +- std dev: [patchfull] 255 ns +- 9 ns -> [patchv2full] 260 ns +- 13 ns: 1.02x slower
array repeat inplace 10: Mean +- std dev: [patchfull] 472 ns +- 8 ns -> [patchv2full] 487 ns +- 22 ns: 1.03x slower
array repeat inplace 1000: Mean +- std dev: [patchfull] 867 ns +- 85 ns -> [patchv2full] 941 ns +- 81 ns: 1.08x slower
[control] create array 1: Mean +- std dev: [patchfull] 224 ns +- 2 ns -> [patchv2full] 226 ns +- 3 ns: 1.01x slower
[control] create array 10: Mean +- std dev: [patchfull] 447 ns +- 3 ns -> [patchv2full] 449 ns +- 3 ns: 1.00x slower
[control] create array 5000: Mean +- std dev: [patchfull] 128 us +- 4 us -> [patchv2full] 127 us +- 3 us: 1.01x faster

Benchmark hidden because not significant (4): bytes(1) repeat 1, array(1) repeat 5000, array repeat inplace 1, [control] create array 1000
``` 
My estimate is the small cases are about 5% slower. Is that acceptable?

One way to reduce two function calls to one would be something like (perhaps with the case `len_src==1` added):
```
void
_PyBytes_Repeat(char* dest, Py_ssize_t len_dest,
    const char* src, Py_ssize_t len_src, bool inplace)
{
    if (len_dest > 0) {
        if (!inplace)
            memcpy(dest, src, len_src);

        Py_ssize_t copied = start_len;
        while (copied < end_len) {
            Py_ssize_t bytes_to_copy = Py_MIN(copied, end_len - copied);
            memcpy(dest + copied, buffer, bytes_to_copy);
            copied += bytes_to_copy;
        }
    }
}
```

Reply:
Thanks for those benchmarks. Out of curiosity, how are you compiling? PGO and LTO can change things.

I'd say the important things to optimize for are code complexity and the linear coefficient, and the O(1) overhead is not too important. Though since the starting-length-1 case is probably the most common, you're probably right that having a memset branch in both functions might be worth it then.

How would you feel about something like this?

```python
def _PyBytes_RepeatInPlace(...):
    if len_dest == 1:
        memset(...)
    while ...:
        ...

def _PyBytes_Repeat(...):
    if len_dest == 0:
        return
    elif len_dest == 1:
        memset(...)
    else:
        memcpy(...)
        _PyBytes_RepeatInPlace(...)
```

I would still prefer non-`static inline` functions though, even if there is a small penalty.

Reply:
The PR has been updated.

* Performance for the cases that this PR was initialy started for (e.g. the inplace repeat of `array.array`) is better than main
* The helper function `_PyBytes_Repeat` is used at 6 places (potentially at 3 more places in https://github.com/python/cpython/pull/32045). If not inlined (this PR) that helps reduce the amount of code.
* Performance is better if we inline the method (branch https://github.com/eendebakpt/cpython/tree/performance/array_inplace_repeat_v3_inline). When compiled with `--enable-optimizations` (so LTO and PGO) and benchmarked with `./python test_array.py --rigorous --min-time .2 -o base.json
` I get:
```
array(3) repeat 0: Mean +- std dev: [patchv3] 162 ns +- 10 ns -> [patchv3inline] 157 ns +- 3 ns: 1.03x faster
array(3) repeat 1: Mean +- std dev: [patchv3] 189 ns +- 4 ns -> [patchv3inline] 183 ns +- 2 ns: 1.03x faster
array(3) repeat 2: Mean +- std dev: [patchv3] 191 ns +- 3 ns -> [patchv3inline] 184 ns +- 9 ns: 1.04x faster
array(3) repeat 10: Mean +- std dev: [patchv3] 222 ns +- 3 ns -> [patchv3inline] 206 ns +- 2 ns: 1.08x faster
array(3) repeat 1000: Mean +- std dev: [patchv3] 1.67 us +- 0.01 us -> [patchv3inline] 1.66 us +- 0.02 us: 1.01x faster
array(1) repeat 10: Mean +- std dev: [patchv3] 101 ns +- 2 ns -> [patchv3inline] 96.0 ns +- 1.6 ns: 1.05x faster
array(1) repeat 1000: Mean +- std dev: [patchv3] 263 ns +- 4 ns -> [patchv3inline] 260 ns +- 4 ns: 1.01x faster
array(1) repeat 5000: Mean +- std dev: [patchv3] 1.40 us +- 0.02 us -> [patchv3inline] 1.40 us +- 0.01 us: 1.00x faster
array(3) repeat inplace 1: Mean +- std dev: [patchv3] 401 ns +- 9 ns -> [patchv3inline] 405 ns +- 12 ns: 1.01x slower
array(3) repeat inplace 10: Mean +- std dev: [patchv3] 462 ns +- 15 ns -> [patchv3inline] 457 ns +- 10 ns: 1.01x faster
int array(3) repeat inplace 0: Mean +- std dev: [patchv3] 364 ns +- 8 ns -> [patchv3inline] 367 ns +- 11 ns: 1.01x slower
int array(3) repeat inplace 1: Mean +- std dev: [patchv3] 347 ns +- 8 ns -> [patchv3inline] 351 ns +- 7 ns: 1.01x slower
bytes(1) repeat 1: Mean +- std dev: [patchv3] 36.6 ns +- 0.4 ns -> [patchv3inline] 37.0 ns +- 0.7 ns: 1.01x slower
bytes(1) repeat 2: Mean +- std dev: [patchv3] 48.2 ns +- 0.4 ns -> [patchv3inline] 49.1 ns +- 1.0 ns: 1.02x slower
bytes(1) repeat 10: Mean +- std dev: [patchv3] 48.0 ns +- 0.6 ns -> [patchv3inline] 48.8 ns +- 0.5 ns: 1.02x slower
bytes(1) repeat 100: Mean +- std dev: [patchv3] 48.5 ns +- 0.7 ns -> [patchv3inline] 49.1 ns +- 1.2 ns: 1.01x slower
bytes(1) repeat 1000: Mean +- std dev: [patchv3] 103 ns +- 2 ns -> [patchv3inline] 102 ns +- 2 ns: 1.01x faster
bytes(7) repeat 1: Mean +- std dev: [patchv3] 36.5 ns +- 0.3 ns -> [patchv3inline] 36.8 ns +- 0.5 ns: 1.01x slower
bytes(7) repeat 2: Mean +- std dev: [patchv3] 50.8 ns +- 0.4 ns -> [patchv3inline] 50.6 ns +- 1.0 ns: 1.00x faster
bytes(7) repeat 10: Mean +- std dev: [patchv3] 59.3 ns +- 0.9 ns -> [patchv3inline] 58.1 ns +- 0.7 ns: 1.02x faster
bytearray(1) repeat 1: Mean +- std dev: [patchv3] 57.9 ns +- 0.8 ns -> [patchv3inline] 55.2 ns +- 0.6 ns: 1.05x faster
bytearray(1) repeat 2: Mean +- std dev: [patchv3] 58.7 ns +- 0.6 ns -> [patchv3inline] 55.9 ns +- 0.9 ns: 1.05x faster
bytearray(1) repeat 10: Mean +- std dev: [patchv3] 58.1 ns +- 0.6 ns -> [patchv3inline] 55.6 ns +- 0.7 ns: 1.04x faster
bytearray(1) repeat 100: Mean +- std dev: [patchv3] 58.2 ns +- 0.5 ns -> [patchv3inline] 56.3 ns +- 1.0 ns: 1.03x faster
bytearray(1) repeat 1000: Mean +- std dev: [patchv3] 72.8 ns +- 1.5 ns -> [patchv3inline] 71.4 ns +- 2.0 ns: 1.02x faster
bytearray(10) repeat 1: Mean +- std dev: [patchv3] 58.2 ns +- 0.9 ns -> [patchv3inline] 55.4 ns +- 0.6 ns: 1.05x faster
bytearray(10) repeat 2: Mean +- std dev: [patchv3] 61.3 ns +- 1.1 ns -> [patchv3inline] 58.3 ns +- 0.7 ns: 1.05x faster
bytearray(10) repeat 10: Mean +- std dev: [patchv3] 69.5 ns +- 1.2 ns -> [patchv3inline] 67.5 ns +- 0.8 ns: 1.03x faster
bytearray(10) repeat 1000: Mean +- std dev: [patchv3] 231 ns +- 6 ns -> [patchv3inline] 229 ns +- 2 ns: 1.01x faster
[control] create array 1: Mean +- std dev: [patchv3] 220 ns +- 4 ns -> [patchv3inline] 221 ns +- 8 ns: 1.01x slower
[control] create array 1000: Mean +- std dev: [patchv3] 25.5 us +- 0.6 us -> [patchv3inline] 25.2 us +- 0.4 us: 1.01x faster
[control] create array 5000: Mean +- std dev: [patchv3] 126 us +- 1 us -> [patchv3inline] 125 us +- 4 us: 1.01x faster

Benchmark hidden because not significant (11): array(3) repeat 5000, array(1) repeat 0, array(1) repeat 1, array(3) repeat inplace 0, array(3) repeat inplace 1000, int array(3) repeat inplace 10, int array(3) repeat inplace 1000, bytes(7) repeat 100, bytes(7) repeat 1000, bytearray(10) repeat 100, [control] create array 10

Geometric mean: 1.01x faster
```

I will try to get a benchmark with `pyperformance`.